### PR TITLE
Add KRestartSamplerCustom + other cleanups

### PR DESCRIPTION
**Pull request by blepping**
this is a semi-draft pull since it's not really tested well as yet. mostly want to see if you're interested in it at all.

this pull:

1. adds a `KRestartSamplerCustom` node that can take a `SAMPLER` input like `SamplerCustom`. it doesn't take sigmas since scheduling seems to require special handling. (of course, there's no protection against using weird samplers with this so some may not work very well or at all)
2. i was able to eliminate the global variables for the schedule/sampler type and steps (but the monkeypatch for `ProgressBar` remains - not sure if there's a better way)
3. comfy must have removed the `t_to_sigma` etc functions that the DDIM scheduler depends on at some point, so i was able to fix that.
4. the list of schedulers the nodes provide are now based on the scheduler map, since it will just immediately crash if you choose one that's not in the list anyway since it attempts to index the scheduler map by key unconditionally. i.e. you could choose SGM before but there was no item in the map so it would always just crash.

i'm not really sure the denoised output/output part is correct, it seems like it's working.

anyway, very open to comments/feedback.

### CODE REVIEW

**Comment by blepping** [comment]
setting denoise actually doesn't work right now, so marking as draft. i will fix it soon.

**Comment by ssitu** [comment]
Thanks for the fixes! I haven't been keeping up with the comfyui updates. What is the denoised_output for?

Is this ready for merge? I've run my old test workflow and everything seems to be working despite some minor differences in the output, which I'm sure is fine since the algorithm is unchanged.

**Review by ssitu — APPROVED** [review]

**Comment by ssitu** [comment]
> these changes allow specifying t_min and t_max using the existing sigmas format, or a negative integer (interpreted as the 0-1000 timestep) or a string like "10%" for percentages. for example, [3,2,"45.5%", "65%"],[4,1,-380,-550],[4,5,0.59,1.09],[3,2,0.06,0.30],[3,1,0.30,0.59]

I am all for making it more user-friendly, and I think the percentages are a good addition. The integers for timesteps is good too, but the user will have to know the range of timesteps that their model is trained on IIRC, so it's the same problem with sigmas but with an easier form to look at. I agree that working with sigmas is terrible, and I definitely would use the percentages as a user.

>  it actually took me a long time to even figure out what t_min and t_max were since it sounded like timesteps.

Haha, sorry about the `t_min` and `t_max` thing. I was trying to keep the notation used in the paper, but I probably should've renamed them to reference sigma in some way.

> hopefully! i am a reasonably competent Python developer and getting a little familiar with ComfyUI's API but i definitely don't know the magical diffusion/restart sampling/etc math well. not really selling my pull here, but i should probably say you shouldn't trust that i know what i'm doing with that part.

I can't say that I know everything about diffusion models either, so I can't guarantee that the way I implemented the algorithm is correct. Since there's no changes to the main sampling part, I suppose that as long as the calculated sampling schedules are similar to before, then it should be fine. I can verify that when I can.

**Comment by blepping** [comment]
i made some changes to the README just down. how does that look, do you see anything that seems like incorrect information? i think the visual example will help people like me actually understand what Restart is doing (and what the segment definitions actually mean). figuring out what a good setting is and stuff like sigmas/percentages is another problem, but at least i think it is a step in the right direction.

>  but the user will have to know the range of timesteps that their model is trained on IIRC

i'm not sure if comfy's conception of timesteps is normal but his code (for discrete sampling at least) just looks like:

```python
    def percent_to_sigma(self, percent):
        if percent <= 0.0:
            return 999999999.9
        if percent >= 1.0:
            return 0.0
        percent = 1.0 - percent
        return self.sigma(torch.tensor(percent * 999.0)).item()
```

so the timestep is basically the inverse of the percentage with a multiplier. 

> Haha, sorry about the t_min and t_max thing.

no apology necessary, i'm just happy restart sampling was available!

> I can't say that I know everything about diffusion models either, so I can't guarantee that the way I implemented the algorithm is correct. 

i'm sure you know a lot more than i do. :)

anyway, unless you think something needs to be changed i think this is good to go now!

**Comment by ssitu** [comment]
Thank you for the changes! I'll go ahead and close those issues.

> i'll be keeping an eye on the issue for a while to see if people run into any problems related to me changes. please also let me know if you find any!

I appreciate the help, I'll make sure to let you know if I do!
